### PR TITLE
Lsp rename feature

### PR DESCRIPTION
**Pull request by Marenz**
Still a WIP, but for all the curious ppl out there.

### CODE REVIEW

File: libsolidity/lsp/RenameSymbol.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+		)
+			m_outer.m_locations.emplace_back(symbolAlias.location);
+}
+
+void RenameSymbol::Visitor::endVisit(MemberAccess const& _node)
+{
+	if (
+		m_outer.m_symbolName == _node.memberName() &&
+		*m_outer.m_declarationToRename == *_node.annotation().referencedDeclaration
+   )
```
Comment (stackenbotten) [inline libsolidity/lsp/RenameSymbol.cpp]: Coding style error

File: libsolidity/lsp/RenameSymbol.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		m_outer.m_locations.emplace_back(_node.memberLocation());
+}
+
+void RenameSymbol::Visitor::endVisit(Identifier const& _node)
+{
+	if (
+		m_outer.m_symbolName == _node.name() &&
+		*m_outer.m_declarationToRename == *_node.annotation().referencedDeclaration
+   )
```
Comment (stackenbotten) [inline libsolidity/lsp/RenameSymbol.cpp]: Coding style error

File: libsolidity/lsp/RenameSymbol.cpp (hunk 1 above)
Comment (stackenbotten) [inline libsolidity/lsp/RenameSymbol.cpp]: Coding style error

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `f476abc00f38520499f11fe01c31078c862ae4e4`:
```
Coding style error:
libsolidity/lsp/RenameSymbol.cpp:182:   )

```
Please check that your changes are working as intended.

File: libsolidity/lsp/RenameSymbol.cpp (hunk 1 above)
Comment (stackenbotten) [inline libsolidity/lsp/RenameSymbol.cpp]: Coding style error

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `54188e6f7de4c154a3249acda20228327170f7c7`:
```
Coding style error:
libsolidity/lsp/RenameSymbol.cpp:36:namespace  {

```
Please check that your changes are working as intended.

File: libsolidity/lsp/RenameSymbol.cpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+#include <memory>
+#include <string>
+#include <vector>
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+using namespace solidity::lsp;
+using namespace std;
+
+namespace  {
```
Comment (stackenbotten) [inline libsolidity/lsp/RenameSymbol.cpp]: Coding style error

File: libsolidity/lsp/RenameSymbol.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+		Json::Value edit = Json::objectValue;
+		edit["range"] = toRange(*i);
+		edit["newText"] = newName;
+
+		// Record changes for the client
+		edits.append(edit);
+		if (i + 1 == m_locations.rend() || (i + 1)->sourceName != i->sourceName)
+		{
+			reply["changes"][uri] = edits;
+
```
Comment (matheusaaguiar) [inline libsolidity/lsp/RenameSymbol.cpp]: Could this blank line be removed? It seems unnecessary to me.

**Review by matheusaaguiar — DISMISSED** [review]
It looks good to me. Only thing that bugged me was that blank line, but maybe you have a good reason for it.

File: libsolidity/lsp/RenameSymbol.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/lsp/RenameSymbol.cpp]: I had it simply so that the body is not all sticking together and easier to read
Removed it now.

Comment (matheusaaguiar) [reply libsolidity/lsp/RenameSymbol.cpp]: I imagined that, but only two lines should be ok :)

**Review by matheusaaguiar — APPROVED** [review]

**Pull request merged**